src/verifier/verifier.rs

```diff
@@ -162,79 +225,225 @@ impl WhammVisitor<Option<DataType>> for TypeChecker<'_> {
         // type check body
 
         self.table.enter_named_scope(&function.name.name);
-        self.visit_block(&function.body);
+        let mut check_ret_type = self.visit_block(&function.body);
         let _ = self.table.exit_scope();
-
-        // return type
-        todo!();
+        if check_ret_type.is_none() {
+            check_ret_type = Some(DataType::Tuple { ty_info: vec![] });
+        }
+        //figure out how to deal with void functions (return type is ())
+        if check_ret_type != function.return_ty {
+            self.err.type_check_error(
+                false,
+                format!(
+                    "The function signature for '{}' returns '{:?}', but the body returns '{:?}'",
+                    function.name.name, function.return_ty, check_ret_type
+                ),
+                &function.name.loc.clone().map(|l| l.line_col),
+            );
+        }
+        //return the type of the fn
+        function.return_ty.clone()
     }
 
     fn visit_block(&mut self, block: &Block) -> Option<DataType> {
-        // TODO: finish user def function type checking
-        for stmt in &block.stmts {
-            self.visit_stmt(stmt);
+        let mut ret_type = None;
+        let num_statements = &block.stmts.len();
+        let start_of_range: usize;
+        for i in 0..*num_statements {
+            let temp = self.visit_stmt(&block.stmts[i]);
+            if temp.is_some() && ret_type.is_none() {
+                ret_type = temp;
+            } else if ret_type.is_some() {
+                start_of_range = i;
+                //get the span for the first statement to the last one
+                let loc = Location::from(
+                    &block.stmts[start_of_range].loc().clone().unwrap().line_col,
+                    &block.stmts[*num_statements - 1]
+                        .loc()
+                        .clone()
+                        .unwrap()
+                        .line_col,
+                    None,
+                );
+                self.err.add_typecheck_warn(
+                    "Unreachable code detected, these statement(s) will not be executed"
+                        .to_string(),
+                    Some(loc.line_col),
+                );
+                return ret_type;
+            }
         }
-        todo!()
+        //add a check for return statement type matching the function return type if provided
+        ret_type
     }
 
     fn visit_stmt(&mut self, stmt: &Statement) -> Option<DataType> {
-        match stmt {
-            Statement::Assign { var_id, expr, .. } => {
-                // change type in symbol table?
-                let lhs_loc = var_id.loc().clone().unwrap();
-                let rhs_loc = expr.loc().clone().unwrap();
-                let lhs_ty_op = self.visit_expr(var_id);
-                let rhs_ty_op = self.visit_expr(expr);
+        if self.in_script_global {
```

Comment:
Factor out common code between if/else blocks (e.g. match bodies for Stmt::Decl and Stmt::Assign)